File: src/li/klass/fhem/appwidget/AppWidgetDataHolder.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2012, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLICLICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 */

package li.klass.fhem.appwidget;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.os.ResultReceiver;
import android.util.Log;
import android.widget.RemoteViews;
import li.klass.fhem.appwidget.view.WidgetType;
import li.klass.fhem.appwidget.view.widget.AppWidgetView;
import li.klass.fhem.constants.Actions;
import li.klass.fhem.constants.BundleExtraKeys;
import li.klass.fhem.constants.ResultCodes;
import li.klass.fhem.domain.core.Device;
import li.klass.fhem.service.room.RoomListService;
import li.klass.fhem.util.ApplicationProperties;

import java.util.Set;

import static li.klass.fhem.util.SharedPreferencesUtil.getSharedPreferences;
import static li.klass.fhem.util.SharedPreferencesUtil.getSharedPreferencesEditor;

public class AppWidgetDataHolder {
    public static final AppWidgetDataHolder INSTANCE = new AppWidgetDataHolder();
    private static final String preferenceName = AppWidgetDataHolder.class.getName();
    private String SAVE_SEPARATOR = "#";

    private AppWidgetDataHolder() {}

    public void updateAllWidgets(final Context context) {
        new AsyncTask<String, String, String>() {
            @Override
            protected String doInBackground(String... voids) {
                Set<String> appWidgetIds = getSharedPreferences(preferenceName).getAll().keySet();
                AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);

                for (String appWidgetId : appWidgetIds) {
                    updateWidgetInCurrentThread(appWidgetManager, context, Integer.parseInt(appWidgetId));
                }
                return null;
            }
        }.doInBackground("");
    }

    public void updateWidget(final AppWidgetManager appWidgetManager, final Context context, final int appWidgetId) {
        new AsyncTask<String, String, String>() {
            @Override
            protected String doInBackground(String... voids) {
                updateWidgetInCurrentThread(appWidgetManager, context, Integer.parseInt(String.valueOf(appWidgetId)));
                return null;
            }
        }.doInBackground("");
    }

    private void updateWidgetInCurrentThread(final AppWidgetManager appWidgetManager, final Context context, final int appWidgetId) {
        final WidgetConfiguration widgetConfiguration = getWidgetConfiguration(appWidgetId);
        if (widgetConfiguration == null) {
            Log.d(AppWidgetDataHolder.class.getName(), "cannot find widget for id " + appWidgetId);
            deleteWidget(context, appWidgetId);
            return;
        }
        final AppWidgetView widgetView = widgetConfiguration.widgetType.widgetView;

        boolean doRemoteWidgetUpdates = ApplicationProperties.INSTANCE.getBooleanSharedPreference("prefWidgetRemoteUpdate", true);
        long updatePeriod = doRemoteWidgetUpdates ? widgetView.updateInterval() : RoomListService.NEVER_UPDATE_PERIOD;
        scheduleUpdateIntent(context, widgetConfiguration, false);

        Intent deviceIntent = new Intent(Actions.GET_DEVICE_FOR_NAME);
        deviceIntent.putExtra(BundleExtraKeys.DEVICE_NAME, widgetConfiguration.deviceName);
        deviceIntent.putExtra(BundleExtraKeys.UPDATE_PERIOD, updatePeriod);
        deviceIntent.putExtra(BundleExtraKeys.RESULT_RECEIVER, new ResultReceiver(new Handler()) {
            @Override
            protected void onReceiveResult(int resultCode, Bundle resultData) {
                if (resultCode == ResultCodes.SUCCESS) {
                    Device device = (Device) resultData.get(BundleExtraKeys.DEVICE);
                    if (device == null) {
                        Log.d(AppWidgetDataHolder.class.getName(), "cannot find device " + widgetConfiguration.deviceName);
                        return;
                    }

                    RemoteViews content = widgetView.createView(context, device, widgetConfiguration);

                    appWidgetManager.updateAppWidget(appWidgetId, content);
                }
            }
        });
        context.startService(deviceIntent);
    }

    public void deleteWidget(Context context, int appWidgetId) {
        SharedPreferences.Editor editor = getSharedPreferencesEditor(preferenceName);
        editor.remove(String.valueOf(appWidgetId));

        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);

        PendingIntent updatePendingIntent = updatePendingIndentForWidgetId(context, appWidgetId);
        alarmManager.cancel(updatePendingIntent);
    }

    public void saveWidgetConfigurationToPreferences(Context context, WidgetConfiguration widgetConfiguration) {
        SharedPreferences.Editor edit = getSharedPreferencesEditor(preferenceName);
        String value = widgetConfiguration.widgetId + SAVE_SEPARATOR
                + widgetConfiguration.deviceName + SAVE_SEPARATOR
                + widgetConfiguration.widgetType.name();
        edit.putString(String.valueOf(widgetConfiguration.widgetId), value);
        edit.commit();

        scheduleUpdateIntent(context, widgetConfiguration, true);
    }

    private void scheduleUpdateIntent(Context context, WidgetConfiguration widgetConfiguration, boolean updateNow) {
        long updateInterval = widgetConfiguration.widgetType.widgetView.updateInterval();
        if (updateInterval > 0) {
            AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
            PendingIntent sender = updatePendingIndentForWidgetId(context, widgetConfiguration.widgetId);
            long now = System.currentTimeMillis();
            long firstRun = updateNow ? now : now + updateInterval;

            alarmManager.setRepeating(AlarmManager.RTC, firstRun, updateInterval, sender);
        }
    }

    private PendingIntent updatePendingIndentForWidgetId(Context context, int widgetId) {
        Intent updateIntent = new Intent(Actions.WIDGET_UPDATE);
        updateIntent.putExtra(BundleExtraKeys.APP_WIDGET_ID, widgetId);

        return PendingIntent.getBroadcast(context, widgetId * (-1),
                updateIntent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    private WidgetConfiguration getWidgetConfiguration(int widgetId) {
        SharedPreferences sharedPreferences = getSharedPreferences(preferenceName);
        String value = sharedPreferences.getString(String.valueOf(widgetId), null);
        if (value == null) return null;

        String[] parts = value.split(SAVE_SEPARATOR);
        if (parts.length != 3) return null;

        WidgetType widgetType = getWidgetTypeFromName(parts[2]);
        if (widgetType == null) {
            return null;
        }

        return new WidgetConfiguration(Integer.valueOf(parts[0]), parts[1], widgetType);
    }

    private WidgetType getWidgetTypeFromName(String widgetTypeName) {
        try {
            return WidgetType.valueOf(widgetTypeName);
        } catch (Exception e) {
            Log.e(AppWidgetDataHolder.class.getName(), "cannot find widget type for name " + widgetTypeName, e);
            return null;
        }
    }
}

File: src/main/java/li/klass/fhem/appwidget/update/AppWidgetDataHolder.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.appwidget.update

import android.app.AlarmManager
import android.app.PendingIntent
import android.appwidget.AppWidgetHost
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.util.Log
import com.google.common.base.Optional
import li.klass.fhem.appwidget.ui.widget.base.AppWidgetView
import li.klass.fhem.constants.Actions
import li.klass.fhem.constants.BundleExtraKeys.ALLOW_REMOTE_UPDATES
import li.klass.fhem.constants.BundleExtraKeys.APP_WIDGET_ID
import li.klass.fhem.update.backend.RoomListService
import li.klass.fhem.util.ApplicationProperties
import li.klass.fhem.util.NetworkState
import li.klass.fhem.util.preferences.SharedPreferencesService
import org.slf4j.LoggerFactory
import java.lang.Integer.parseInt
import javax.inject.Inject

class AppWidgetDataHolder @Inject
constructor() {

    @Inject
    lateinit var applicationProperties: ApplicationProperties
    @Inject
    lateinit var sharedPreferencesService: SharedPreferencesService

    fun updateAllWidgets(context: Context, allowRemoteUpdate: Boolean) {
        val appWidgetIds = getAllAppWidgetIds(context)
        for (appWidgetId in appWidgetIds) {
            context.startService(getRedrawWidgetIntent(context, parseInt(appWidgetId), allowRemoteUpdate))
        }
    }

    fun getAllAppWidgetIds(context: Context): Set<String> {
        val sharedPreferences = getSavedPreferences(context)
        val allEntries = sharedPreferences.all!!

        return allEntries.keys
    }

    private fun getRedrawWidgetIntent(context: Context, appWidgetId: Int, allowRemoteUpdate: Boolean): Intent {
        return Intent(Actions.REDRAW_WIDGET)
                .setClass(context, AppWidgetUpdateIntentService::class.java)
                .putExtra(APP_WIDGET_ID, appWidgetId)
                .putExtra(ALLOW_REMOTE_UPDATES, allowRemoteUpdate)
    }

    private fun getSavedPreferences(context: Context): SharedPreferences =
            sharedPreferencesService.getPreferences(SAVE_PREFERENCE_NAME, context)


    fun getAppWidgetView(configuration: WidgetConfiguration): AppWidgetView =
            configuration.widgetType.widgetView

    fun deleteWidget(context: Context, appWidgetId: Int) {
        Log.d(AppWidgetDataHolder::class.java.name, String.format("deleting widget for id %d", appWidgetId))

        val preferences = getSavedPreferences(context)
        val key = appWidgetId.toString()
        if (preferences.contains(key)) {
            preferences.edit().remove(key).apply()

            val host = getAppWidgetHost(context)
            host.deleteAppWidgetId(appWidgetId)
        }

        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        cancelUpdating(context, appWidgetId, alarmManager)
    }

    fun getAppWidgetHost(context: Context): AppWidgetHost = AppWidgetHost(context, 0)

    private fun cancelUpdating(context: Context, appWidgetId: Int, alarmManager: AlarmManager) {
        val updatePendingIntent = updatePendingIndentForWidgetId(context, appWidgetId)
        alarmManager.cancel(updatePendingIntent)
    }

    fun scheduleUpdateIntent(context: Context, widgetConfiguration: WidgetConfiguration,
                             updateImmediately: Boolean, widgetUpdateInterval: Long) {
        if (widgetUpdateInterval > 0) {
            LOG.debug(String.format("scheduling widget update %s => %s, updateImmediately=%b ", widgetConfiguration.toString(), (widgetUpdateInterval / 1000).toString() + "s", updateImmediately))

            val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
            val pendingIntent = updatePendingIndentForWidgetId(context, widgetConfiguration.widgetId)
            val now = System.currentTimeMillis()
            val firstRun = if (updateImmediately) now else now + widgetUpdateInterval

            cancelUpdating(context, widgetConfiguration.widgetId, alarmManager)
            if (pendingIntent != null) {
                alarmManager.setRepeating(AlarmManager.RTC, firstRun, widgetUpdateInterval, pendingIntent)
            }
        }
    }

    private fun updatePendingIndentForWidgetId(context: Context, widgetId: Int): PendingIntent? {
        val updateIntent = getRedrawWidgetIntent(context, widgetId, true)

        return PendingIntent.getService(context, widgetId * -1,
                updateIntent, PendingIntent.FLAG_UPDATE_CURRENT)
    }

    fun getConnectionDependentUpdateInterval(context: Context): Long {
        val updateInterval: Long

        if (!NetworkState.isConnected(context)) {
            LOG.debug("getConnectionDependentUpdateInterval - no network connection")
            updateInterval = RoomListService.Companion.NEVER_UPDATE_PERIOD
        } else if (NetworkState.isConnectedMobile(context)) {
            LOG.debug("getConnectionDependentUpdateInterval - mobile connection")
            updateInterval = getWidgetUpdateIntervalFor(WIDGET_UPDATE_INTERVAL_PREFERENCES_KEY_MOBILE).toLong()
        } else {
            LOG.debug("getConnectionDependentUpdateInterval - wlan connection")
            updateInterval = getWidgetUpdateIntervalFor(WIDGET_UPDATE_INTERVAL_PREFERENCES_KEY_WLAN).toLong()
        }

        return updateInterval
    }

    fun getWidgetConfiguration(widgetId: Int, context: Context): Optional<WidgetConfiguration> {
        val sharedPreferences = getSavedPreferences(context)
        val value = sharedPreferences.getString(widgetId.toString(), null)

        if (value == null) {
            return Optional.absent<WidgetConfiguration>()
        } else {
            return Optional.fromNullable(WidgetConfiguration.fromSaveString(value))
        }
    }

    fun saveWidgetConfigurationToPreferences(widgetConfiguration: WidgetConfiguration, context: Context) {
        val edit = sharedPreferencesService.getSharedPreferencesEditor(SAVE_PREFERENCE_NAME, context)
        val value = widgetConfiguration.toSaveString()
        edit.putString(widgetConfiguration.widgetId.toString(), value)
        edit.apply()
    }

    private fun getWidgetUpdateIntervalFor(key: String): Int {
        val value = applicationProperties.getStringSharedPreference(key, "3600")
        val intValue = parseInt(value)
        return intValue * 1000
    }

    companion object {
        private val WIDGET_UPDATE_INTERVAL_PREFERENCES_KEY_WLAN = "WIDGET_UPDATE_INTERVAL_WLAN"
        private val WIDGET_UPDATE_INTERVAL_PREFERENCES_KEY_MOBILE = "WIDGET_UPDATE_INTERVAL_MOBILE"
        val SAVE_PREFERENCE_NAME = AppWidgetDataHolder::class.java.name
        val LOG = LoggerFactory.getLogger(AppWidgetDataHolder::class.java)
    }
}
